[ETH-FLOW V2] - New CLASSIC ETH/WETH Flow [Swap + Wrap + Unwrap]

W3stside opened the pull request:
# Summary

#### _Currently only the new ETH/WETH wrapping, WETH/ETH unwrapping, and classic WETH trading functionalities are available in this PR._

This modifies/changes/improves the ETH/WETH flow in regards to:
1. Wrapping ETH > WETH
2. Unwrapping WETH > ETH
3. Swapping ETH > WETH <> ERC20 by obfuscating some of the wrap + approve logic necessary

### Native ETH flow is NOT available yet

# Testing
**Swapping ETH <> ERC20**
1. choose ETH as sell (or any native token)
2. choose an ERC20 as buy
3. Swap
Notes: you should get the entire wrapping/approving flow (if necessary)

**Wrapping ETH <> WETH**
1. choose ETH as sell (or any native token)
2. choose WETH as buy
3. Wrap

**Unwrapping WETH <> ETH**
1. choose WETH as sell (or any wrapped native token)
2. choose ETH as buy
3. Wrap

# Screenshots
### Non-expertMode
1. No more old eth flow!
<img width="520" alt="image" src="https://user-images.githubusercontent.com/21335563/186468292-d5b84874-da1f-4b77-9a10-cbb6d5a5b504.png">

2. Wrap ETH modal (if necessary)
<img width="630" alt="image" src="https://user-images.githubusercontent.com/21335563/186468434-7278ec8e-c3ee-4269-9fed-b0ffa11adacc.png">

3. Wrap ETH modal: pending transaction
 <img width="581" alt="image" src="https://user-images.githubusercontent.com/21335563/186468611-fa7decae-3062-43f7-aba1-d4d2d7752100.png">

4. Approve WETH
<img width="556" alt="image" src="https://user-images.githubusercontent.com/21335563/186469291-6d0674c5-ec28-47e6-9348-907e36ead6cf.png">

5. Approve WETH: pending transaction
<img width="542" alt="image" src="https://user-images.githubusercontent.com/21335563/186469408-8f2688a4-6c32-4951-8ef1-46165328ce30.png">

6. Confirmed meta-txs before swap
<img width="518" alt="image" src="https://user-images.githubusercontent.com/21335563/186469479-f7204302-391b-486b-8483-997aa11ae2f3.png">

### expertMode
1. Wrap and approve both needed
<img width="567" alt="image" src="https://user-images.githubusercontent.com/21335563/186469778-3c8e2287-f458-40ed-b7a3-68a725eb950d.png">

2. Both operations pending 
<img width="550" alt="image" src="https://user-images.githubusercontent.com/21335563/186469871-5e2ffce7-b6d2-474e-a19b-b021da79fdd3.png">

3. Goes to swap confirmation
<img width="1003" alt="image" src="https://user-images.githubusercontent.com/21335563/186469947-78288c14-74ba-4ccd-a60a-a1bfd8515cdd.png">

### TX failures
##### Note: to test, do as @elena-zh does, set a custom very low gasLimit (22k) and it will fail after some time
<img width="1017" alt="image" src="https://user-images.githubusercontent.com/21335563/186470207-2e354aa3-cd2e-4e25-b561-d5ec2e88dea7.png">

☝️ also applies the same to Approve failures and if BOTH fail during expertMode simultaneous tx signing

### Docs
Figma:
https://www.figma.com/proto/UGqD48GRUCiyav8z1xhhH2/%5BUX-RESEARCH%5D-%5D(W)ETH-Flows-(2022)?page-id=1%3A29&node-id=644%3A3899&viewport=976%2C-2470%2C0.06&scaling=min-zoom&starting-point-node-id=644%3A2927&show-proto-sidebar=1

Spec doc: https://docs.google.com/document/d/1AKZ7G3quQUw1PAbI43q64lMXHqg_BF9d9PQT4ciBaVQ/edit#


### Docs
Figma:
https://www.figma.com/proto/UGqD48GRUCiyav8z1xhhH2/%5BUX-RESEARCH%5D-%5D(W)ETH-Flows-(2022)?page-id=1%3A29&node-id=644%3A3899&viewport=976%2C-2470%2C0.06&scaling=min-zoom&starting-point-node-id=644%3A2927&show-proto-sidebar=1

Spec doc: https://docs.google.com/document/d/1AKZ7G3quQUw1PAbI43q64lMXHqg_BF9d9PQT4ciBaVQ/edit#

josojo reviewed (commented):
PR looks good. I will give it a second go once the logic changes are updated as discussed.

Generally, in this PR there are a lot of if else conditions :( Must have been hard to implement. How do you usually make sure that so many if else conditions are all correct?

W3stside commented on src/custom/pages/Swap/SwapMod.tsx:
```diff
@@ -37,7 +37,10 @@ import SwapHeader from 'components/swap/SwapHeader'
 import { useAllTokens, useCurrency } from 'hooks/Tokens'
-import { ApprovalState /*, useApprovalOptimizedTrade*/, useApproveCallbackFromTrade } from 'hooks/useApproveCallback'
+import {
+  ApprovalState /*, useApprovalOptimizedTrade, OptionalApproveCallbackParams,*/,
```
so this comes from the way we modify Uniswap files. Because we want to keep the original code the same, we comment out stuff we don't use.

That way when we do backmerges we can see what was changed in the diff

Makes it very awkward and messy but it does simplify the merge a ton

W3stside commented on src/custom/components/AccountDetails/Transaction/styled.ts:
```diff
@@ -78,12 +78,12 @@ export const IconType = styled.div`
   width: 100%;
   grid-template-rows: 1fr;
-  grid-template-columns: 80px max-content;
+  grid-template-columns: ${({ renderSimple }) => (renderSimple ? 'auto' : '80px max-content')};
```
good catch! it is obsolete. was originally using props to manipulate/conditionally render but kept it simple in the end and am using CSS. will remove

W3stside replied:
to avoid having to rebase the waterfalls I will apply this in a later PR, it is in my notes

W3stside replied:
#913

W3stside replied:
done in this PR

josojo commented on src/custom/components/swap/EthWethWrap/index.tsx:
```diff
@@ -113,160 +67,448 @@ const ButtonWrapper = styled.div`
+          wrapHash: wrapTx?.hash,
+          approveHash: approveTx?.hash,
+        }))
+        setApproveSubmitted(true)
```
Is this incorrect, if only wrapping was needed?

josojo commented on src/custom/components/swap/EthWethWrap/index.tsx:
```diff
@@ -113,160 +67,448 @@ const ButtonWrapper = styled.div`
+  native,
+  wrapped,
+  nativeInput,
+  isNativeIn,
```
(just a thought: probably not actionable)
To determine your actions, you are handing over a lot of bool variables(nativeInput,isNativeIn,isNativeOut,isWrappedIn,isWrappedOut,) to the EthWethWrap function. 
Maybe it would have been nicer to hand over a flow type enum: 
```
enum ETHWethWrapType {
WrapETH,
UnwrapETH,
WrapETHAndTrade
}
'
and having a helper function that calculates the enum ETHWethWrapType from the bool variables. This could increase the readability of the code. 
But it seems that your way is much easier and also more the normal FE design pattern.

W3stside replied:
I 100% agree that there are too many prop bool flags being sent thru and it could be optimised. I'll take a look at it and consider your suggestion, I like it

W3stside replied:
done [7325ff9](https://github.com/cowprotocol/cowswap/pull/889/commits/7325ff9b3cd419881b83fe0bba79ff5ac0c20809)

josojo reviewed (approved):
Very nice!

I tested the PR thoroughly, and it worked as discussed. 🎉

josojo commented on src/custom/components/swap/EthWethWrap/index.tsx:
```diff
@@ -113,160 +67,600 @@ const ButtonWrapper = styled.div`
+  // and sometimes there can be 2 simultaneous transactions
+  if (isExpertMode) {
+    if (state === EthFlowState.WrapAndApproveNeeded) {
+      header = `${isNative ? 'Wrap ' + nativeSymbol : 'Unwrap ' + wrappedSymbol} and Approve`
```
Does "unwrapping and approve" make sense? In which scenario would a user want to do it?

W3stside replied:
woudn't happen, ure right

josojo reviewed (commented):
Very nice! I tested the PR one more time and it looks all great!

I think the readability of the code improved quite a bit! Very nice. I like that you factored quite a bunch of logic into separate functions

alfetopito commented on src/custom/components/AccountDetails/SimpleAccountDetails.tsx:
```diff
@@ -0,0 +1,50 @@
+  StyledWrapperProps
+
+const SimpleWrapper = styled(Wrapper)<StyledWrapperProps>`
+  ${({ $margin }) => $margin && `margin: ${$margin};`}
```
What happens when `$margin` is falsy?
Is the resulting falsy value included in the output css?

W3stside replied:
falsy in styleds returns nothing, like null or undefined or ''

alfetopito commented on src/custom/components/AccountDetails/SimpleAccountDetails.tsx:
```diff
@@ -0,0 +1,50 @@
+
+  return (
+    <SimpleWrapper {...styleProps}>
+      {!!pendingTransactions.length || !!confirmedTransactions.length ? (
```
Nitpick: you don't really need `!!`, or do you?

W3stside replied:
nope!

W3stside replied:
```suggestion
      {pendingTransactions.length || confirmedTransactions.length ? (
```

alfetopito commented on src/custom/components/swap/EthWethWrap/index.tsx:
```diff
@@ -113,160 +67,600 @@ const ButtonWrapper = styled.div`
+
+      if (type === 'WRAP') {
+        setWrapError(error)
+      }
+      if (type === 'APPROVE') {
```
Nitpick:

```suggestion
      } else if (type === 'APPROVE') {
```

alfetopito replied:
Or even, since it's a required parameter and there are only 2 choices:

```suggestion
      } else {
```

alfetopito replied:
Or even shorter which for sure @anxolin will dislike:

```ts
type === 'WRAP' ? setWrapError(error) : setApproveError(error)
```

W3stside replied:
haha yep was avoiding that (although i prefer it)

elena-zh commented:
I have done some tests in https://github.com/cowprotocol/cowswap/pull/903#issuecomment-1220799942 and metioned issues that are related to the current PR. 
Their numbers are 5, 7, 8.

elena-zh commented:
To add: 
1. the modal looks different than other modals in a mobile view: 
![image](https://user-images.githubusercontent.com/70885163/185657742-000c68fa-91a5-4de5-8a18-89ee7dc76151.png)
compare with 
![image](https://user-images.githubusercontent.com/70885163/185657845-6fd2d745-b7fe-4907-a69f-b5d17f846f90.png)

2. Looks like there are square boarders on the ETH elements in the dark mode 
![image](https://user-images.githubusercontent.com/70885163/185658025-8f5bd7f2-2ae3-4067-a45f-c2d5c7f18684.png)

W3stside commented on src/custom/pages/Swap/SwapMod.tsx:
```diff
@@ -241,8 +252,8 @@ export default function Swap({
-    // if native input !== NATIVE_TOKEN, validation fails
-    nativeInput || parsedAmount,
+    // is native token swap, use the wrapped equivalent as input currency
+    isNativeInSwap ? (nativeInput ?? parsedAmount)?.wrapped : nativeInput ?? parsedAmount,
```
yeah it's true, || makes more sense here

W3stside replied:
```suggestion
    isNativeInSwap ? (nativeInput || parsedAmount)?.wrapped : nativeInput || parsedAmount,
```

W3stside commented:
> To add:
> 
> 1. the modal looks different than other modals in a mobile view:
>    ![image](https://user-images.githubusercontent.com/70885163/185657742-000c68fa-91a5-4de5-8a18-89ee7dc76151.png)
>    compare with
>    ![image](https://user-images.githubusercontent.com/70885163/185657845-6fd2d745-b7fe-4907-a69f-b5d17f846f90.png)
> 2. Looks like there are square boarders on the ETH elements in the dark mode
>    ![image](https://user-images.githubusercontent.com/70885163/185658025-8f5bd7f2-2ae3-4067-a45f-c2d5c7f18684.png)

@fairlighteth what are your thoughts on the design?

W3stside commented:
@elena-zh I am re-requesting your review now - I've addressed the component style changes as well as the issues you outlined in #903 regarding erroring txs: https://github.com/cowprotocol/cowswap/pull/903#issuecomment-1220799942

W3stside commented:
@elena-zh after some internal discussion with the RD team it makes the most sense to review this PR first completely. The Toggle and Banner will be decided on later and thus merged later.

Bringing in the comments you made on #903 here to review:
> Hey @W3stside , cool changes! Still. I need to read and investigate new ETH flow requirements, so here I did tests in an ad-hoc manner. That is why some issues that I will list below might be irrelevant a bit. So here they are:
> 
> 1. New slippage field text is a bit different than the rest fields
>    ![slipp diff](https://user-images.githubusercontent.com/70885163/185648524-80ca3235-47f9-4840-98d3-cf50ec2073d6.jpg)

DONE

> 2. I'm a bit concerned about the 'copy' icon on the message (why I need to copy it/why I can't copy?)
>    ![image](https://user-images.githubusercontent.com/70885163/185648757-01af4e98-6a73-4007-bae8-2aa7dcc6fbe5.png)

DONE

> 3. Auto option is the settings is enabled, I can switch to it when new ETH flow is enabled
>    ![can press on auto](https://user-images.githubusercontent.com/70885163/185649169-8d011d3f-2c4f-434a-87e8-48be98948873.jpg)

Not applicable in this PR

> 4. Can we auto-return a user to the 0.5% slippage when we switch off new ETH flow option? Currently, we leave it on 2% value
>    ![return to auto](https://user-images.githubusercontent.com/70885163/185649361-e31bd3fb-19c2-4633-b3fb-ae038ce66bac.jpg)

Not applicable in this PR but done in #902 

> 5. When I use a classic WETH/ETH flow, and wrap max ETH amount, the app does not show me an additional warning
>    ![warning](https://user-images.githubusercontent.com/70885163/185649642-d39c5fe8-5149-4b75-9f9f-15cbc3bfec2e.jpg)
>    ![missing warning](https://user-images.githubusercontent.com/70885163/185649783-b8c90a8d-5b1a-4815-8960-714ca0950ed9.jpg)

Having issues reproding this, but I am aware of what the issue is mentally. Not super blocking imo

> 6. In addition to @fairlighteth note about disabling the toggle to a not connected user, I would also suggest to disable it when an error is displayed (insuff liquidity or others)
>    ![errors](https://user-images.githubusercontent.com/70885163/185650062-ea8d2a1e-7d16-4bd9-84df-59549d5366e8.jpg)

DONE

> 7. View details link on Approve token is shifted to top in reduced/mobile screens
>    ![view details at top](https://user-images.githubusercontent.com/70885163/185650242-91dd6a3e-c3b1-4c9a-a823-2f06cc62f636.jpg)

This is aligned with what the Orders panel does (it's for mobile users) but I can add a little padding or sth

> 8. When Approve token TX fails, Swap button is enabled and sends an order request (which will fail). So I would show 'Approve' button again instead of 'Swap'
>    ![failed -- swap](https://user-images.githubusercontent.com/70885163/185650637-7084e2bf-bf55-420a-93f6-10e8315598bc.jpg)
>    ![result of swap](https://user-images.githubusercontent.com/70885163/185650640-0ac16b31-7a17-4d9e-a022-807a362991fd.jpg)
> 
> The same for failed wrap ![image](https://user-images.githubusercontent.com/70885163/185655740-e369d485-db21-4600-a300-5268a5893ba6.png)
> 

DONE - ooof good catch. This one was tough

> 9. I can get 'Wrap Wrapped Ether button: Select a new flow option, fill in ETH/DAI option, press on Swap, close approve modal, change ETH token to WETH (see the video: https://watch.screencastify.com/v/4itijIdFLoEs9g92BcxL)
>    ![image](https://user-images.githubusercontent.com/70885163/185651212-2564f71b-9b2d-4710-991e-abf8ed63c862.png)

Not applicable here but also done in #903 

> 10. I believe, that it is not yet implemented, but it is a bit unclear to me why I see Approve WETH when I select ETH in the Sell field. Besides, when I run the transaction, WETH token is displayed in the Account modal
>     ![not classic - not clear why approve weth](https://user-images.githubusercontent.com/70885163/185651856-8b89bfa8-cd0f-4144-90dc-c123560e28d2.jpg)
> 
> ![place order - weth](https://user-images.githubusercontent.com/70885163/185651466-20dd2e2b-328a-45d8-aa27-6b0294362906.jpg)
> 
> Thanks!

You're correct, it's cause it's not implemented yet.

W3stside commented:
@josojo @fairlighteth tagging u again here.

@josojo please check the latest flow - i've addressed multiple requested changes from @elena-zh in #903 here and this would be the PR to merge into dev & prod.

@fairlighteth please review the styles in the eth-flow modals as I've addressed the comments u and elena made in #903 here

josojo reviewed (commented):
I really like the code and I tested it one more time. Works great.

-> As discussed in sync some logic could go into their own files, but this is the only real feedback left from my side.
Otherwise, looks great

elena-zh commented:
Hey @W3stside , great job!
Some bugs you hate from my side: 
1. UI: Wrap/Unwrap container looks like not 100% filled with a color
![container](https://user-images.githubusercontent.com/70885163/186662625-fb3b73a2-c274-4652-bc15-3a6499337545.jpg)
![not filled in](https://user-images.githubusercontent.com/70885163/186662635-6396928d-b3f4-4856-be8f-cf3610cb4916.jpg)
2. I noticed that fields may overlap each other in a mobile view
![image](https://user-images.githubusercontent.com/70885163/186662935-616b1887-b1aa-4455-bf91-2b26d58f7e4a.png)
3. Regular mode: once I press on wrap/unwrap, waiting modal appear and it covers additional wrap/unwrap modal. When close the waiting modal, and press on the button on the wrap/unwrap modal, it sends the 2nd request to the wallet. See the video: https://watch.screencastify.com/v/sqVNPOQiWSKP6qJB7drZ
Then when I sign a wrap/unwrap request in the wallet, I do not see TX progress in the new modals as it is auto-closed.
4. Expert mode: (a bit not sure about it) no pending TX details on wrap/unwrap/approve when a single TX is executed (not both such wrap and approve)
5. Edge case: a user can do a partial approval of a token. I mean, I wrap and fill in amount of 5ETH, but I approve 3 ETH. After a successful approval, I'm navigated to the Swap modal and I'm allowed to proceed. Hwever, the TX will fail in this case. So it would be great to handle this case and call approve token button once again if WETH token allowance is not enough for a trade
![Screenshot_1](https://user-images.githubusercontent.com/70885163/186664569-95863d8b-95f9-4ea0-968a-cd5857e7b199.jpg)
6. I have not noticed wrap+approve functionality for a regular mode. I have to wrap a token (and it is wrapped), then I'm navigated to the Swap form, and a request fails as WETH token is not approved. 
7. Expert mode: When wrap and approve TXs fail, I request these actions once again, an empty button appears on the modal
![empty button again](https://user-images.githubusercontent.com/70885163/186665215-4f595b8a-7fed-4042-8d7f-ae1fe3d2e25d.jpg)

Thanks!

josojo reviewed (approved):
LGTM. great

elena-zh commented:
> > When I use a classic WETH/ETH flow, and wrap max ETH amount, the app does not show me an additional warning
> 
> Having issues reproding this, but I am aware of what the issue is mentally. Not super blocking imo

From my perspective it is a missing functionality. 
To reproduce (in Prod) fill ie ETH amount close to 100% ETH balance in the wallet 
![image](https://user-images.githubusercontent.com/70885163/186683575-97bfe1e7-c767-483c-a4f7-8e97912a5f9f.png)

elena-zh commented:
All the rest issues reported in #903 task,  cases 1-2 are fixed and verified. 
All the rest cases I will review when it is fully implemented and ready for me to review. 

Thanks!

W3stside commented:
> > > When I use a classic WETH/ETH flow, and wrap max ETH amount, the app does not show me an additional warning
> > 
> > 
> > Having issues reproding this, but I am aware of what the issue is mentally. Not super blocking imo
> 
> From my perspective it is a missing functionality. To reproduce (in Prod) fill ie ETH amount close to 100% ETH balance in the wallet ![image](https://user-images.githubusercontent.com/70885163/186683575-97bfe1e7-c767-483c-a4f7-8e97912a5f9f.png)

it should be fine now in the latest build, could you give it a look please? thanks!

W3stside commented:
> Hey @W3stside , great job! Some bugs you hate from my side:

😁 😁 😁 😁 

> 3. Regular mode: once I press on wrap/unwrap, waiting modal appear and it covers additional wrap/unwrap modal. When close the waiting modal, and press on the button on the wrap/unwrap modal, it sends the 2nd request to the wallet. See the video: https://watch.screencastify.com/v/sqVNPOQiWSKP6qJB7drZ
>    Then when I sign a wrap/unwrap request in the wallet, I do not see TX progress in the new modals as it is auto-closed.

### This also happens in production, so i'm thinking we don't change anymore non-prod things here

> 4. Expert mode: (a bit not sure about it) no pending TX details on wrap/unwrap/approve when a single TX is executed (not both such wrap and approve)

### This also happens in production, so i'm thinking we don't change anymore non-prod things here

> 5. Edge case: a user can do a partial approval of a token. I mean, I wrap and fill in amount of 5ETH, but I approve 3 WETH. After a successful approval, I'm navigated to the Swap modal and I'm allowed to proceed. However, the TX will fail in this case. So it would be great to handle this case and call approve token button once again if WETH token allowance is not enough for a trade
>    ![Screenshot_1](https://user-images.githubusercontent.com/70885163/186664569-95863d8b-95f9-4ea0-968a-cd5857e7b199.jpg)

### Hmmm, shit. You're right, I'll see what I can do for this..

> 6. I have not noticed wrap+approve functionality for a regular mode. I have to wrap a token (and it is wrapped), then I'm navigated to the Swap form, and a request fails as WETH token is not approved.

### Hmmm, I've never seen this as I coded specifically for both to show up synchronously. Let me see if something happened

> 7. Expert mode: When wrap and approve TXs fail, I request these actions once again, an empty button appears on the modal
>    ![empty button again](https://user-images.githubusercontent.com/70885163/186665215-4f595b8a-7fed-4042-8d7f-ae1fe3d2e25d.jpg)

### Ok will take a look, I think I've seen this too

> 
> Thanks!

elena-zh commented:
Hey @W3stside !

> > From my perspective it is a missing functionality. To reproduce (in Prod) fill ie ETH amount close to 100% ETH balance in the wallet ![image](https://user-images.githubusercontent.com/70885163/186683575-97bfe1e7-c767-483c-a4f7-8e97912a5f9f.png)
> 
> it should be fine now in the latest build, could you give it a look please? thanks!

Great that you have added this warning!
However, it would be great to make it look better in a dark mode
![colors](https://user-images.githubusercontent.com/70885163/187230719-53d580e3-005e-4c14-8460-869209a4d10e.jpg)

elena-zh commented:
> 
> > 3. Regular mode: once I press on wrap/unwrap, waiting modal appear and it covers additional wrap/unwrap modal. When close the waiting modal, and press on the button on the wrap/unwrap modal, it sends the 2nd request to the wallet. See the video: https://watch.screencastify.com/v/sqVNPOQiWSKP6qJB7drZ
> >    Then when I sign a wrap/unwrap request in the wallet, I do not see TX progress in the new modals as it is auto-closed.
> 
> ### This also happens in production, so i'm thinking we don't change anymore non-prod things here
> 

Yes, it works this somehow this way in Prod, but there is no underlaying additional modal that can call the transaction again when I close Waiting modal. See the [video](https://watch.screencastify.com/v/VBNd1xZZn5EyxW7epwsU): the 1st part (up to 20 sec) shows the current PR, then I switch to Prod version.

So here is my point is to remove the underlying modal. 
OR
Show this modal as an intermediate state, and then all the Waiting modal.

W3stside commented:
> > > 3. Regular mode: once I press on wrap/unwrap, waiting modal appear and it covers additional wrap/unwrap modal. When close the waiting modal, and press on the button on the wrap/unwrap modal, it sends the 2nd request to the wallet. See the video: https://watch.screencastify.com/v/sqVNPOQiWSKP6qJB7drZ
> > >    Then when I sign a wrap/unwrap request in the wallet, I do not see TX progress in the new modals as it is auto-closed.
> > 
> > 
> > ### This also happens in production, so i'm thinking we don't change anymore non-prod things here
> 
> Yes, it works this somehow this way in Prod, but there is no underlaying additional modal that can call the transaction again when I close Waiting modal. See the [video](https://watch.screencastify.com/v/VBNd1xZZn5EyxW7epwsU): the 1st part (up to 20 sec) shows the current PR, then I switch to Prod version.
> 
> So here is my point is to remove the underlying modal. OR Show this modal as an intermediate state, and then all the Waiting modal.

I see i see, makes sense, I'll make a change! :)
